src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -704,28 +711,29 @@ private BoundLambda ReallyBindForErrorRecovery()
             // and bind.
 
             return
-                GuessBestBoundLambda(_bindingCache.Values)
-                ?? GuessBestBoundLambda(_returnInferenceCache.Values)
+                GuessBestBoundLambda(_bindingCache)
+                ?? GuessBestBoundLambda(_returnInferenceCache)
                 ?? ReallyInferReturnType(null, ImmutableArray<TypeSymbol>.Empty, ImmutableArray<RefKind>.Empty);
         }
 
-        private static BoundLambda GuessBestBoundLambda(ICollection<BoundLambda> candidates)
+        private static BoundLambda GuessBestBoundLambda<T>(ImmutableDictionary<T, BoundLambda> candidates)
         {
             switch (candidates.Count)
             {
                 case 0:
                     return null;
                 case 1:
-                    return candidates.First();
+                    return candidates.First().Value;
                 default:
                     // Prefer candidates with fewer diagnostics.
-                    IEnumerable<BoundLambda> minDiagnosticsGroup = candidates.GroupBy(lambda => lambda.Diagnostics.Length).OrderBy(group => group.Key).First();
+                    IEnumerable<KeyValuePair<T, BoundLambda>> minDiagnosticsGroup = candidates.GroupBy(lambda => lambda.Value.Diagnostics.Length).OrderBy(group => group.Key).First();
 
                     // If multiple candidates have the same number of diagnostics, order them by delegate type name.
                     // It's not great, but it should be stable.
                     return minDiagnosticsGroup
-                        .OrderBy(lambda => GetLambdaSortString(lambda.Symbol))
-                        .FirstOrDefault();
+                        .OrderBy(lambda => GetLambdaSortString(lambda.Value.Symbol))
+                        .FirstOrDefault()
```

Comment:
>FirstOrDefault [](start = 25, length = 14)

This probably should be ```First()```, it look like we don't expect the sequence to be empty.